In [82]:
import pandas as pd
import matplotlib.pyplot as plt
import pickle
import seaborn as sns

In [91]:
salary_df = pd.read_csv('proj_data/data_for_salary_prediction.csv')
salary_df.drop(['Unnamed: 0'] ,axis = 1, inplace = True)
salary_df.head(50)

Job Title  \
0                             Data Scientist   
1                  Healthcare Data Scientist   
2                             Data Scientist   
3                             Data Scientist   
4                             Data Scientist   
5                             Data Scientist   
6                             Data Scientist   
7                             Data Scientist   
8                         Research Scientist   
9                             Data Scientist   
10                            Data Scientist   
11                            Data Scientist   
12         Staff Data Scientist - Technology   
13                              Data Analyst   
14                            Data Scientist   
15                           Data Engineer I   
16                   Scientist I/II, Biology   
17                   Customer Data Scientist   
18    Data Scientist - Health Data Analytics   
19                            Data Scientist   
20                            Data Scientist   
21  Senior Data Scientist / Machine Learning   
22             Data Scientist - Quantitative   
23                            Data Scientist   
24                            Data Scientist   
25                            Data Scientist   
26             Digital Health Data Scientist   
27                            Data Scientist   
28                    Associate Data Analyst   
29                   Clinical Data Scientist   
30                            Data Scientist   
31                            Data Scientist   
32                            Data Scientist   
33  Data Scientist / Machine Learning Expert   
34                            Data Scientist   
35                            Data Scientist   
36                          Web Data Analyst   
37                            Data Scientist   
38                            Data Scientist   
39                     Senior Data Scientist   
40                             Data Engineer   
41                              Data Analyst   
42                            Data Scientist   
43                             Data Engineer   
44   Data Scientist - Algorithms & Inference   
45                                 Scientist   
46                     Senior Data Scientist   
47                      Data Science Analyst   
48                            Data Scientist   
49                       Lead Data Scientist   

                         Salary Estimate  \
0             $53K-$91K (Glassdoor est.)   
1            $63K-$112K (Glassdoor est.)   
2             $80K-$90K (Glassdoor est.)   
3             $56K-$97K (Glassdoor est.)   
4            $86K-$143K (Glassdoor est.)   
5            $71K-$119K (Glassdoor est.)   
6             $54K-$93K (Glassdoor est.)   
7            $86K-$142K (Glassdoor est.)   
8             $38K-$84K (Glassdoor est.)   
9           $120K-$160K (Glassdoor est.)   
10          $126K-$201K (Glassdoor est.)   
11           $64K-$106K (Glassdoor est.)   
12          $106K-$172K (Glassdoor est.)   
13            $46K-$85K (Glassdoor est.)   
14           $83K-$144K (Glassdoor est.)   
15          $102K-$190K (Glassdoor est.)   
16           $67K-$137K (Glassdoor est.)   
17          $118K-$189K (Glassdoor est.)   
18          $110K-$175K (Glassdoor est.)   
19           $64K-$111K (Glassdoor est.)   
20           $81K-$130K (Glassdoor est.)   
21           $73K-$119K (Glassdoor est.)   
22           $86K-$139K (Glassdoor est.)   
23           $63K-$105K (Glassdoor est.)   
24                                    -1   
25          $109K-$177K (Glassdoor est.)   
26           $63K-$110K (Glassdoor est.)   
27           $75K-$124K (Glassdoor est.)   
28            $34K-$61K (Glassdoor est.)   
29           $63K-$105K (Glassdoor est.)   
30            $80K-$90K (Glassdoor est.)   
31            $56K-$97K (Glassdoor est.)   
32           $72K-$120K (Glassdoor est.)   
33           $86K-$143K (Glassdoor est.)   
34           $93K-$149K (Glassdoor est.)   
35

# First,we have to clean the Salary estimate

In [84]:
salary_negve = salary_df[salary_df['Salary Estimate'] != '-1']
len(salary_negve)

742

There are 742 instances where '-1' is the salary estimate

#### Now, we have to do some spliting

In [85]:
salary = salary_df['Salary Estimate'].apply(lambda x: x.split('(')[0])
salary

0       $53K-$91K 
1      $63K-$112K 
2       $80K-$90K 
3       $56K-$97K 
4      $86K-$143K 
          ...     
951    $72K-$133K 
952     $56K-$91K 
953    $95K-$160K 
954             -1
955    $61K-$126K 
Name: Salary Estimate, Length: 956, dtype: object

##### Remove the K and dollar-sign from the salary-estimate 

In [86]:
salary = salary.apply(lambda x: x.replace('K' , '').replace('$' , ''))
salary

0       53-91 
1      63-112 
2       80-90 
3       56-97 
4      86-143 
        ...   
951    72-133 
952     56-91 
953    95-160 
954         -1
955    61-126 
Name: Salary Estimate, Length: 956, dtype: object

#### Now, look at the hourly salary

In [87]:
salary_df['Hourly Salary'] = salary_df['Salary Estimate'].apply(lambda x: 1 if 'per hour' in x.lower() else 0)
salary_df.loc[salary_df['Hourly Salary'] == 1].head()

Job Title  \
197                              Medical Lab Scientist   
209                MED TECH/LAB SCIENTIST - LABORATORY   
240                       Medical Laboratory Scientist   
247          MED TECH/LAB SCIENTIST- SOUTH COASTAL LAB   
257  Medical Technologist / Clinical Laboratory Sci...   

                      Salary Estimate  \
197  $17-$24 Per Hour(Glassdoor est.)   
209  $21-$34 Per Hour(Glassdoor est.)   
240  $18-$25 Per Hour(Glassdoor est.)   
247  $21-$34 Per Hour(Glassdoor est.)   
257  $15-$25 Per Hour(Glassdoor est.)   

                                       Job Description  Rating  \
197  Overview\n\nAbout\n\n\nTower Health is a stron...     3.6   
209  Evening: 3P-1130P or Night Shift 11P-730A\n\nI...     3.6   
240  Description\nMedical Laboratory Scientist - Te...     4.0   
247  Day Shift: 7A-330P. Holidays and every other w...     3.6   
257  The incumbent performs a broad range of labora...     3.4   

                                          Company Name          Location  \
197                                  Tower Health\n3.6  West Reading, PA   
209                              Beebe Healthcare\n3.6         Lewes, DE   
240                 Texas Health Huguley Hospital\n4.0      Burleson, TX   
247                              Beebe Healthcare\n3.6     Millville, DE   
257  Veterans Affairs, Veterans Health Administrati...   New Orleans, LA   

       Headquarters                     Size  Founded       Type of ownership  \
197     Reading, PA  5001 to 10000 employees     2017  Nonprofit Organization   
209       Lewes, DE   1001 to 5000 employees     1935  Nonprofit Organization   
240   Arlington, TX   1001 to 5000 employees     1977                Hospital   
247       Lewes, DE   1001 to 5000 employees     1935  Nonprofit Organization   
257  Washington, DC         10000+ employees     1930              Government   

                             Industry       Sector  \
197  Health Care Services & Hospitals  Health Care   
209  Health Care Services & Hospitals  Health Care   
240  Health Care Services & Hospitals  Health Care   
247  Health Care Services & Hospitals  Health Care   
257                  Federal Agencies   Government   

                        Revenue Competitors  Hourly Salary  
197    Unknown / Non-Applicable          -1              1  
209  $100 to $500 million (USD)          -1              1  
240   $50 to $100 million (USD)          -1              1  
247  $100 to $500 million (USD)          -1              1  
257    Unknown / Non-Applicable          -1              1

In [88]:
# Remove 'per hour' from the salary estimate
salary_df['Salary Estimate'] = salary_df['Salary Estimate'].apply(lambda x: x.lower().replace('per hour' , ''))
salary_df.head()

Job Title              Salary Estimate  \
0             Data Scientist   $53k-$91k (glassdoor est.)   
1  Healthcare Data Scientist  $63k-$112k (glassdoor est.)   
2             Data Scientist   $80k-$90k (glassdoor est.)   
3             Data Scientist   $56k-$97k (glassdoor est.)   
4             Data Scientist  $86k-$143k (glassdoor est.)   

                                     Job Description  Rating  \
0  Data Scientist\nLocation: Albuquerque, NM\nEdu...     3.8   
1  What You Will Do:\n\nI. General Summary\n\nThe...     3.4   
2  KnowBe4, Inc. is a high growth information sec...     4.8   
3  *Organization and Job ID**\nJob ID: 310709\n\n...     3.8   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

                                 Company Name         Location  \
0                      Tecolote Research\n3.8  Albuquerque, NM   
1  University of Maryland Medical System\n3.4    Linthicum, MD   
2                                KnowBe4\n4.8   Clearwater, FL   
3                                   PNNL\n3.8     Richland, WA   
4                     Affinity Solutions\n2.9     New York, NY   

     Headquarters                    Size  Founded   Type of ownership  \
0      Goleta, CA   501 to 1000 employees     1973   Company - Private   
1   Baltimore, MD        10000+ employees     1984  Other Organization   
2  Clearwater, FL   501 to 1000 employees     2010   Company - Private   
3    Richland, WA  1001 to 5000 employees     1965          Government   
4    New York, NY     51 to 200 employees     1998   Company - Private   

                           Industry                        Sector  \
0               Aerospace & Defense           Aerospace & Defense   
1  Health Care Services & Hospitals                   Health Care   
2                 Security Services             Business Services   
3                            Energy  Oil, Gas, Energy & Utilities   
4           Advertising & Marketing             Business Services   

                            Revenue  \
0         $50 to $100 million (USD)   
1            $2 to $5 billion (USD)   
2        $100 to $500 million (USD)   
3  $500 million to $1 billion (USD)   
4          Unknown / Non-Applicable   

                                         Competitors  Hourly Salary  
0                                                 -1              0  
1                                                 -1              0  
2                                                 -1              0  
3  Oak Ridge National Laboratory, National Renewa...              0  
4               Commerce Signals, Cardlytics, Yodlee              0

##### There is also some Employer provided salaries, make a column for them

In [89]:
salary_df['Employer_provided_salary'] = salary_df['Salary Estimate'].apply(lambda x : 1 if 'employer provided salary:' in x.lower() else 0)
salary_df.loc[salary_df['Employer_provided_salary'] == 1].head()

Job Title  \
48                                      Data Scientist   
161                                      Data Engineer   
208  Principal Data Scientist with over 10 years ex...   
331  Principal Data Scientist with over 10 years ex...   
377                               Data Operations Lead   

                          Salary Estimate  \
48   employer provided salary:$150k-$160k   
161  employer provided salary:$120k-$145k   
208  employer provided salary:$200k-$250k   
331  employer provided salary:$200k-$250k   
377    employer provided salary:$85k-$90k   

                                       Job Description  Rating  \
48   BPA Services, LLC is seeking a Computer/Data S...     5.0   
161  Location: Tampa, FL\nTitle: Data Engineer\nTS/...     5.0   
208  Position Title: Principal Data Scientist\nLoca...    -1.0   
331  Position Title: Principal Data Scientist\nLoca...    -1.0   
377  Data Operations Lead\nLocation: Flexible thoug...    -1.0   

          Company Name           Location       Headquarters  \
48   BPA Services\n5.0     Washington, DC     Alexandria, VA   
161   Gridiron IT\n5.0          Tampa, FL         Reston, VA   
208  CA-One Tech Cloud  San Francisco, CA        Fremont, CA   
331  CA-One Tech Cloud  San Francisco, CA        Fremont, CA   
377               Muso  San Francisco, CA  San Francisco, CA   

                     Size  Founded       Type of ownership  \
48                Unknown       -1       Company - Private   
161   51 to 200 employees     2017       Company - Private   
208   51 to 200 employees     2017       Company - Private   
331   51 to 200 employees     2017       Company - Private   
377  201 to 500 employees       -1  Nonprofit Organization   

                                    Industry                  Sector  \
48   Enterprise Software & Network Solutions  Information Technology   
161                              IT Services  Information Technology   
208                              IT Services  Information Technology   
331                              IT Services  Information Technology   
377                                       -1                      -1   

                      Revenue Competitors  Hourly Salary  \
48   Unknown / Non-Applicable          -1              0   
161  Unknown / Non-Applicable          -1              0   
208   $5 to $10 million (USD)          -1              0   
331   $5 to $10 million (USD)          -1              0   
377  Unknown / Non-Applicable          -1              0   

     Employer_provided_salary  
48                          1  
161                         1  
208                         1  
331                         1  
377                         1

In [93]:
# Remove 'Employer Provided Salary' from the salary Estimate
salary_df['Salary Estimate'] = salary_df['Salary Estimate'].apply(lambda x: x.lower().replace('employer provided salary:' , ''))
salary_df.head(50)

Job Title               Salary Estimate  \
0                             Data Scientist    $53k-$91k (glassdoor est.)   
1                  Healthcare Data Scientist   $63k-$112k (glassdoor est.)   
2                             Data Scientist    $80k-$90k (glassdoor est.)   
3                             Data Scientist    $56k-$97k (glassdoor est.)   
4                             Data Scientist   $86k-$143k (glassdoor est.)   
5                             Data Scientist   $71k-$119k (glassdoor est.)   
6                             Data Scientist    $54k-$93k (glassdoor est.)   
7                             Data Scientist   $86k-$142k (glassdoor est.)   
8                         Research Scientist    $38k-$84k (glassdoor est.)   
9                             Data Scientist  $120k-$160k (glassdoor est.)   
10                            Data Scientist  $126k-$201k (glassdoor est.)   
11                            Data Scientist   $64k-$106k (glassdoor est.)   
12         Staff Data Scientist - Technology  $106k-$172k (glassdoor est.)   
13                              Data Analyst    $46k-$85k (glassdoor est.)   
14                            Data Scientist   $83k-$144k (glassdoor est.)   
15                           Data Engineer I  $102k-$190k (glassdoor est.)   
16                   Scientist I/II, Biology   $67k-$137k (glassdoor est.)   
17                   Customer Data Scientist  $118k-$189k (glassdoor est.)   
18    Data Scientist - Health Data Analytics  $110k-$175k (glassdoor est.)   
19                            Data Scientist   $64k-$111k (glassdoor est.)   
20                            Data Scientist   $81k-$130k (glassdoor est.)   
21  Senior Data Scientist / Machine Learning   $73k-$119k (glassdoor est.)   
22             Data Scientist - Quantitative   $86k-$139k (glassdoor est.)   
23                            Data Scientist   $63k-$105k (glassdoor est.)   
24                            Data Scientist                            -1   
25                            Data Scientist  $109k-$177k (glassdoor est.)   
26             Digital Health Data Scientist   $63k-$110k (glassdoor est.)   
27                            Data Scientist   $75k-$124k (glassdoor est.)   
28                    Associate Data Analyst    $34k-$61k (glassdoor est.)   
29                   Clinical Data Scientist   $63k-$105k (glassdoor est.)   
30                            Data Scientist    $80k-$90k (glassdoor est.)   
31                            Data Scientist    $56k-$97k (glassdoor est.)   
32                            Data Scientist   $72k-$120k (glassdoor est.)   
33  Data Scientist / Machine Learning Expert   $86k-$143k (glassdoor est.)   
34                            Data Scientist   $93k-$149k (glassdoor est.)   
35                            Data Scientist   $85k-$140k (glassdoor est.)   
36                          Web Data Analyst   $77k-$135k (glassdoor est.)   
37                            Data Scientist   $82k-$132k (glassdoor est.)   
38                            Data Scientist   $83k-$137k (glassdoor est.)   
39                     Senior Data Scientist  $115k-$180k (glassdoor est.)   
40                             Data Engineer   $74k-$138k (glassdoor est.)   
41                              Data Analyst   $64k-$112k (glassdoor est.)   
42                            Data Scientist                            -1   
43                             Data Engineer   $68k-$129k (glassdoor est.)   
44   Data Scientist - Algorithms & Inference  $110k-$175k (glassdoor est.)   
45                                 Scientist   $52k-$113k (glassdoor est.)   
46                     Senior Data Scientist    $110k-$150k(employer est.)   
47                      Data Science Analyst                            -1   
48                            Data Scientist                   $150k-$160k   
49                       Lead Data Scientist  $158k-$211k (glassdoor est.)   

                                      Job Description  R

#### Minimum  and Maximum Salary

In [112]:
salary_df['Minimum salary'] = salary.apply(lambda x: int(x.split('-')[0]))
salary_df

ValueError: invalid literal for int() with base 10: ''

In [102]:
salary_df['Minimum salary'].dtype

dtype('O')